vault/logical_system_raft.go

```diff
@@ -176,23 +176,35 @@ func (b *SystemBackend) handleRaftRemovePeerUpdate() framework.OperationFunc {
 
 func (b *SystemBackend) handleRaftBootstrapChallengeWrite() framework.OperationFunc {
 	return func(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-		_, ok := b.Core.underlyingPhysical.(*raft.RaftBackend)
+		raftStorage, ok := b.Core.underlyingPhysical.(*raft.RaftBackend)
 		if !ok {
 			return logical.ErrorResponse("raft storage is not in use"), logical.ErrInvalidRequest
 		}
 
+		leader, err := raftStorage.IsLeader(ctx)
+		if err != nil {
+			return nil, err
+		}
+		if !leader {
+			return logical.ErrorResponse("not the raft leader node"), logical.ErrInvalidRequest
+		}
+
 		serverID := d.Get("server_id").(string)
 		if len(serverID) == 0 {
 			return logical.ErrorResponse("no server id provided"), logical.ErrInvalidRequest
 		}
 
-		uuid, err := uuid.GenerateRandomBytes(16)
-		if err != nil {
-			return nil, err
+		answer, ok := b.Core.pendingRaftPeers[serverID]
```

Comment:
We certainly need to iterate over the leader infos even in case of shamir. We will need this logic. Moreover, I saw "concurrent map writes" panic while testing. To solve it, I'm now not writing to the map every time a challenge is created. Instead we only write for the first time.